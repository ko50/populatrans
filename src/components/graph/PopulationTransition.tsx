import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
} from "chart.js";
import { Line } from "react-chartjs-2";

import { TransitionList } from "models/transition";
import styles from "styles/components/graph/population_transition.module.scss";
import { filterGraphLabel } from "controllers/populationTransitionGraph/filterGraphLabel";
import { conversionPopulationTransition } from "controllers/populationTransitionGraph/conversionGraphData";

type Props = {
  transitionList?: TransitionList;
};

export const PopulationTransitionGraph: React.FC<Props> = ({
  transitionList,
}) => {
  if (transitionList === undefined)
    return <div className={styles.wrapper}>Loading</div>;

  ChartJS.register(
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend
  );

  const options = {
    responsive: true,
    plugins: {
      legend: {
        position: "right" as const,
      },
      title: {
        display: false,
      },
    },
  };

  const labels = filterGraphLabel(transitionList);
  const data = conversionPopulationTransition(labels, transitionList);

  return (
    <div className={styles.wrapper}>
      <div className={styles.population}>人口数</div>
      <div className={styles.year}>年度</div>
      <div className={styles.graph}>
        <Line options={options} data={data}></Line>
      </div>
    </div>
  );
};
