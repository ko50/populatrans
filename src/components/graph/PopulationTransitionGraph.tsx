import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
} from "chart.js";
import { Line } from "react-chartjs-2";

import { TransitionList } from "models/transition";
import styles from "styles/components/graph.module.scss";

type Props = {
  transition: TransitionList;
};

export const PopulationTransitionGraph: React.FC<Props> = ({ transition }) => {
  ChartJS.register(
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend
  );

  const options = {
    responsive: true,
    plugins: {
      legend: {
        position: "right" as const,
      },
      title: {
        display: false,
      },
    },
  };

  const labelSet: Set<number> = new Set();
  const _labels = transition.transitions.map((t) =>
    t.populations.map((p) => p.year)
  );
  _labels.forEach((array) => {
    array.forEach((label) => {
      const consistent = _labels.every((array2) => array2.includes(label));
      if (consistent) labelSet.add(label);
    });
  });
  const labels = Array.from(labelSet.values()).sort();

  const data = {
    labels,
    datasets: transition.transitions.map((t) => {
      return {
        label: t.pref,
        data: t.populations
          .filter((p) => labels.includes(p.year))
          .map((p) => p.value),
        borderColor: "rgb(255, 99, 132)",
        backgroundColor: "rgba(255, 99, 132, 0.5)",
      };
    }),
  };

  return (
    <div className={styles.wrapper}>
      <div className={styles.population}>人口数</div>
      <div className={styles.year}>年度</div>
      <div className={styles.graph}>
        <Line options={options} data={data}></Line>
      </div>
    </div>
  );
};
