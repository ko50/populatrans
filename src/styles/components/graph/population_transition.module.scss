@use "styles/utils/mq.scss";
@use "styles/utils/colors.scss";
@use "styles/utils/fonts.scss";

.wrapper {
  height: 33rem;
  width: 100%;
  position: relative;
  padding: 1rem;
  background-color: colors.$spring;
}

.navigation {
  @extend .wrapper;

  font-family: fonts.$kosugi;
  color: colors.$carbon;
  display: flex;
  align-items: center;
  justify-content: center;
}

.population {
  position: absolute;
  top: 1rem;
  left: 1rem;

  @include mq.mq("sm") {
    font-size: 0.8rem;
  }
}

.year {
  position: absolute;
  right: 3rem;
  bottom: 1.5rem;

  @include mq.mq("md") {
    right: 5.5rem;
    bottom: 3rem;
  }

  @include mq.mq("sm") {
    right: 2rem;
    bottom: 1rem;
    font-size: 0.8rem;
  }
}

.graph {
  padding: 2rem;
  min-height: 30rem;
  width: 100%;

  @include mq.mq("sm") {
    padding: 0 {
      top: 1.5rem;
    }
  }
}

.copyright {
  font-size: 0.7rem;
  text-align: right;
  color: colors.$carbon;
  font-family: fonts.$kosugi;
}
