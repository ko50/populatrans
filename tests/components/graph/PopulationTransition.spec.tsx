import React from "react";
import renderer from "react-test-renderer";
import { PopulationTransitionGraph } from "components/graph/PopulationTransition";

describe("population transition graph component", () => {
  it("should render correctly", () => {
    const tree = renderer
      .create(
        <PopulationTransitionGraph
          transitionList={{ boundaryYear: 2021, transitions: [] }}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render correctly with actual data", () => {
    const tree = renderer
      .create(
        <PopulationTransitionGraph
          transitionList={{
            boundaryYear: 2021,
            transitions: [
              {
                pref: "茨城県",
                populations: [
                  {
                    year: 2014,
                    value: 200,
                  },
                  {
                    year: 2015,
                    value: 200,
                  },
                  {
                    year: 2016,
                    value: 1600,
                  },
                  {
                    year: 2017,
                    value: 0,
                  },
                  {
                    year: 2018,
                    value: 500,
                  },
                  {
                    year: 2019,
                    value: 4500,
                  },
                  {
                    year: 2020,
                    value: 6000,
                  },
                  {
                    year: 2021,
                    value: 2000,
                  },
                ],
              },
            ],
          }}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
