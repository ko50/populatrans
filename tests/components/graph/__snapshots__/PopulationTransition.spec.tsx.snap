// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`population transition graph component should render correctly 1`] = `
<div
  className="navigation"
>
  表示するデータがありません。都道府県を選択してください
</div>
`;

exports[`population transition graph component should render correctly with actual data 1`] = `
<div
  className="wrapper"
>
  <div
    className="population"
  >
    人口数
  </div>
  <div
    className="year"
  >
    年度
  </div>
  <div
    className="graph"
  >
    <canvas
      height={150}
      role="img"
      width={300}
    />
  </div>
  <div
    className="copyright"
  >
    RESAS (地域経済分析システム) を加工して作成
  </div>
</div>
`;
